lib/ansible/plugins/inventory/tower.py

```diff
@@ -0,0 +1,201 @@
+
+try:
+    import requests
+    from requests.auth import HTTPBasicAuth
+except ImportError:
+    raise AnsibleError('You have to install python requests library to use tower inventory plugin.')
+
+
+class InventoryModule(BaseInventoryPlugin):
+    NAME = 'tower'
+    # Stays backward compatible with tower inventory script.
+    # If the user supplies 'tower_inventory' as path, the plugin will read from environment variables.
+    read_all_configs_from_env_var = False
+
+    def read_tower_inventory(self, tower_host, tower_user, tower_pass, inventory, license_type, verify_ssl=True):
+        if not re.match('(?:http|https)://', tower_host):
+            tower_host = "https://{tower_host}".format(tower_host=tower_host)
+        inventory_id = inventory.replace('/', '')
+        inventory_url = "/api/v2/inventories/{inv_id}/script/?hostvars=1&towervars=1&all=1".format(inv_id=inventory_id)
+        inventory_url = urljoin(tower_host, inventory_url)
+        config_url = urljoin(tower_host, "/api/v2/config/")
+        try:
+            if license_type != "open":
+                config_response = requests.get(config_url,
+                                               auth=HTTPBasicAuth(tower_user, tower_pass),
+                                               verify=verify_ssl)
+                if config_response.ok:
+                    source_type = config_response.json()['license_info']['license_type']
+                    if not source_type == license_type:
+                        raise AnsibleParserError("Tower server licenses must match: source: {source} local: {local}".
+                                                 format(source=source_type, local=license_type))
+                else:
+                    raise AnsibleParserError("Failed to validate the license of the remote Tower: {response}".
+                                             format(response=config_response))
+
+            response = requests.get(inventory_url,
+                                    auth=HTTPBasicAuth(tower_user, tower_pass),
+                                    verify=verify_ssl)
+            if not response.ok:
+                # If the GET /api/v2/inventories/N/script is not HTTP 200, print the error code
+                raise AnsibleParserError("Connection to remote host failed: {response}".format(response=response))
+            try:
+                # Attempt to parse JSON
+                return response.json()
+            except (ValueError, TypeError) as e:
+                # If the JSON parse fails, print the ValueError
+                raise AnsibleParserError("Failed to parse json from host: {err}".format(err=e))
+        except requests.ConnectionError as e:
+            raise AnsibleParserError("Connection to remote host failed: {err}".format(err=e))
+
+    def verify_file(self, path):
+        if path.endswith('tower_inventory'):
+            self.read_all_configs_from_env_var = True
+            return True
```

Comment:
I agree with the other comments that this is hard to follow, this should look for vars from options first, and then check for the env var if the option is not provided.

Reply:
Yeah this is tricky. I think if the user want to read everything from environment variables, they still needs to provide a `path` in the command line to trigger the plugin, so I put tower_inventory as an option if they do not want to use any config file. I will switch to other implements if there is a better way.

Reply:
For the first-pass, I wouldn't balk at requiring a file with `plugin: tower` in it.

The question I'm more interested in is "do other plugins offer this in this type of format?"

Reply:
I _think_ all other inventory plugins currently use a file. Since most employ their own verify_file and don't always call `super(InventoryModule, self).verify_file(path)` there is kind of a dichotomy though.

Reply:
no, host_list and advance_host_list use an 'inventory string' , which is why they dont call super version of verify_file as 'its not a file'

Reply:
Ah, okay.

Maybe if non-file inventory is to be supported, a more transparent way of doing this would be to remove self.read_all_configs_from_env_var, which does sound like it's indicating an either/or scenario and then put in parse() something like...
```
if os.path.isfile(path):
    try:
        config = self.loader.load_from_file(path)
    except Exception as e:
        raise AnsibleParserError(to_native(e))
    self.set_options(direct=config)
```